StRoot/StFwdTrackMaker/include/Tracker/CriteriaKeeper.h

```diff
@@ -0,0 +1,73 @@
+#ifndef CRITERIA_KEEPER_H
+#define CRITERIA_KEEPER_H
+
+#include "Criteria/Criteria.h"
+
+class CriteriaKeeper : public KiTrack::ICriterion {
+
+  public:
+    CriteriaKeeper(KiTrack::ICriterion *child) {
+        mChild = child;
+        values.clear();
+
+        _name = mChild->getName();
+        _type = mChild->getType();
+    }
+
+    ~CriteriaKeeper() {
+        if (mChild)
+            delete mChild;
+        mChild = nullptr;
+    }
+
+    virtual bool areCompatible(KiTrack::Segment *parent, KiTrack::Segment *child) {
+        bool result = mChild->areCompatible(parent, child);
+
+        // capture the primary result of the criteria
+        float value = mChild->getMapOfValues()[mChild->getName()];
+        bool same_track = false;
+        int track_id = -1;
+
+        // two hit criteria
+        if ((parent->getHits().size() == 1) && (child->getHits().size() == 1)) {
+            KiTrack::IHit *a = parent->getHits()[0];
+            KiTrack::IHit *b = child->getHits()[0];
+
+            same_track = (static_cast<FwdHit *>(a)->_tid == static_cast<FwdHit *>(b)->_tid && static_cast<FwdHit *>(a)->_tid != 0);
+            if (same_track)
+                track_id = static_cast<FwdHit *>(a)->_tid;
+        }
+
+        // three hit criteria (two two-segments)
+        if ((parent->getHits().size() == 2) && (child->getHits().size() == 2)) {
+            KiTrack::IHit *a = child->getHits()[0];
+            KiTrack::IHit *b = child->getHits()[1];
+            KiTrack::IHit *c = parent->getHits()[1];
+
+            same_track = (static_cast<FwdHit *>(a)->_tid == static_cast<FwdHit *>(b)->_tid && static_cast<FwdHit *>(b)->_tid == static_cast<FwdHit *>(c)->_tid && static_cast<FwdHit *>(a)->_tid != 0);
+            if (same_track)
+                track_id = static_cast<FwdHit *>(a)->_tid;
+        }
```

Comment:
The fact that you have so many casts in the code usually indicative of a less than optimal design. The problem may not be with your code but with the library you are using. In fact, I just looked up how you use the Seed_t container in KiTracks. It does have an awful interface `KiTracks::SegementBuilder::SegmentBuilder(  std::map< int , std::vector< IHit* > > map_sector_hits );`

One option is just to modify KiTracks, e.g. add a template. Another workaround would be to use in your code `std::vector<FwdHit>` but create a  vector of pointers before passing to KiTracks.

As I think it is beyond this review I'll leave it up to you.